[WIP] Test (and more fixes) for duplicate indices with concat

ivirshup opened the pull request:
- [ ] closes #31308, #36263
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Follow up to #38654, currently WIP.

It's a work in progress because it turns out union and intersection act differently for different index types. `intersection` seems mostly normal, except for `IntervalIndex` #38743), while `union` is sometimes a set union (for `Index`) and sometimes keeps duplicates (for many `Index` subclasses, though order matters sometimes).

This effects `concat` in `get_objs_combined_axis`. Once I figured out where the problems were coming from, I figured I could maybe side step this by adding checks to `get_objs_combined_axis` for equality and uniqueness. All checks pass `reshape` on my machine, but I'm done working on this for the day, so I'll rely on CI to do the checks on the rest of the codebase.

------------------------

This test could be expanded with:

* More types of indexes (there is no `object` index currently)
* Checks for commutativity, i.e. results should have set-equal columns regardless of order
* Interactions with other arguments (`sort`, `axis`?)

TODO:

- [ ] Move test for `concat` where duplicates don't error
- [ ] Decide whether error should be `DuplicateLabelErrors`
- [ ] Have error report which duplicates are used
- [ ] Check that error reports which duplicates are used

ivirshup committed:
Initial test and fix (WIP)

arw2019 commented on pandas/core/indexes/api.py:
```diff
@@ -89,6 +89,19 @@ def get_objs_combined_axis(
+                copy = False
+            except TypeError:
+                pass
+        if copy:
```
shouldn't this come before `sort`? Presumably if sort=True, copy=True we want to copy first then sort?

ivirshup replied:
Doesn't `sort_values` always return a copy?

arw2019 replied:
So it does

Maybe OT but should we then warn if copy=false and sort=true (or do we do that already?)

ivirshup replied:
I'd say no. The only cases where it's possible to not copy are: when all indices are references to the same object or (now) when all objects have the same value. Otherwise a new value is created, so why even copy?

This is also often not super user visible, since the only time this argument is even passed is in `concat`, where the documentation is clear `copy=False` is "best effort" instead of a strictly enforced rule.

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -445,6 +446,26 @@ def test_concat_ordered_dict(self):
+    index_unique = index_maker(k=4)
+    index_non_unique = index_unique[[0, 0, 1, 2, 3]]
+    with pytest.raises(InvalidIndexError):
+        _ = pd.concat(
```
make the frames outside of this function

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -89,6 +89,19 @@ def get_objs_combined_axis(
     Index
     """
     obs_idxes = [obj._get_axis(axis) for obj in objs]
+    if all_indexes_same(obs_idxes):
```
this is not the place to do at all. should be in concat itself.

ideally we want a nice error message if this happens (and the indices are not equal).

ivirshup replied:
I agree I need to add the error message, but I do think this function is where the error should be thrown.

Every function which uses `get_objs_combined_axis` throws an error if there are duplicate indices. Here is a demonstration of this from pandas 1.1.5:

<details>

```python
import numpy as np
import pandas as pd
import pandas._testing as tm

def list_of_series_constructor(*args):
    return pd.DataFrame(list(args))

def crosstab(*args):
    return pd.crosstab(*args)

def concat_series(*args):
    return pd.concat(list(args), axis=1)
    
def concat_dfs(*args):
    return pd.concat([pd.DataFrame(x) for x in args], axis=1)

def duplicated_indices_errors(func, index_makers):
    records = []
    for i, index_maker in enumerate(index_makers):
        rec = {"index_maker": index_maker.__name__}
        uniq = index_maker(k=4)
        non_uniq = uniq[[0, 0, 1, 2]]

        s_uniq = pd.Series(np.ones(len(uniq)), index=uniq)
        s_non_uniq = pd.Series(np.ones(len(non_uniq)), index=non_uniq)

        try:
            res = func(s_uniq, s_non_uniq)
        except Exception as e:
            rec["error_type"] = str(type(e))
            rec["error_message"] = str(e)
        records.append(rec)
    return records

funcs = [list_of_series_constructor, crosstab, concat_series, concat_dfs]

index_makers = [
    tm.makeStringIndex,
    tm.makeIntIndex,
    tm.makeUIntIndex,
    tm.makeFloatIndex,
    tm.makeDateIndex,
    tm.makeTimedeltaIndex,
    tm.makePeriodIndex,
    tm.makeMultiIndex,
    tm.makeBoolIndex,
]

pd.concat(
    {
        f.__name__: pd.DataFrame(duplicated_indices_errors(f, index_makers)) 
        for f in funcs
    }
).to_markdown()
```

|                                   | index_maker        | error_type                                | error_message                                            |
|:----------------------------------|:-------------------|:------------------------------------------|:---------------------------------------------------------|
| ('list_of_series_constructor', 0) | makeStringIndex    | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 1) | makeIntIndex       | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 2) | makeUIntIndex      | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 3) | makeFloatIndex     | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 4) | makeDateIndex      | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 5) | makeTimedeltaIndex | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 6) | makePeriodIndex    | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 7) | makeMultiIndex     | <class 'ValueError'>                      | Reindexing only valid with uniquely valued Index objects |
| ('list_of_series_constructor', 8) | makeBoolIndex      | <class 'pandas.errors.InvalidIndexError'> | Reindexing only valid with uniquely valued Index objects |
| ('crosstab', 0)                   | makeStringIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 1)                   | makeIntIndex       | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 2)                   | makeUIntIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 3)                   | makeFloatIndex     | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 4)                   | makeDateIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 5)                   | makeTimedeltaIndex | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 6)                   | makePeriodIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('crosstab', 7)                   | makeMultiIndex     | <class 'ValueError'>                      | cannot handle a non-unique multi-index!                  |
| ('crosstab', 8)                   | makeBoolIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 0)              | makeStringIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 1)              | makeIntIndex       | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 2)              | makeUIntIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 3)              | makeFloatIndex     | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 4)              | makeDateIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 5)              | makeTimedeltaIndex | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 6)              | makePeriodIndex    | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_series', 7)              | makeMultiIndex     | <class 'ValueError'>                      | cannot handle a non-unique multi-index!                  |
| ('concat_series', 8)              | makeBoolIndex      | <class 'ValueError'>                      | cannot reindex from a duplicate axis                     |
| ('concat_dfs', 0)                 | makeStringIndex    | <class 'ValueError'>                      | Shape of passed values is (5, 2), indices imply (4, 2)   |
| ('concat_dfs', 1)                 | makeIntIndex       | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 2)                 | makeUIntIndex      | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 3)                 | makeFloatIndex     | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 4)                 | makeDateIndex      | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 5)                 | makeTimedeltaIndex | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 6)                 | makePeriodIndex    | <class 'ValueError'>                      | Shape of passed values is (7, 2), indices imply (5, 2)   |
| ('concat_dfs', 7)                 | makeMultiIndex     | <class 'ValueError'>                      | cannot handle a non-unique multi-index!                  |
| ('concat_dfs', 8)                 | makeBoolIndex      | <class 'ValueError'>                      | Shape of passed values is (4, 2), indices imply (2, 2)   |

</details>


Instead of just giving concat a better error message, all these methods could get better (and more consistent) error messages. I think it's important for the rules about unions and intersections of indices to be defined in these more internal methods so behaviour isn't defined (and reimplemented) per top level function.

Additionally, this function essentially returns nonsense for unions when there are duplicated indices since different `Index` types have different definitions for unions with duplicates, so I think it's appropriate to throw errors here instead of passing on those nonsense results.

This has also made me realize there is one case where it's okay for there to be duplicates, which is when indices contain duplicates, but the intersection would not.

ivirshup replied:
> This has also made me realize there is one case where it's okay for there to be duplicates, which is when indices contain duplicates, but the intersection would not.

This got a bit more complicated. `idx.get_indexer(keys)` throws an error if `idx` is non-unique, regardless of whether `keys` have unique indices in `idx`.

Interestingly enough, it does not throw an error if the `idx` is non-unique, but `keys` is empty (which is the case I had initially seen).

Examples using pandas 1.1.5:

```python
pd.crosstab(
    pd.Series(np.arange(4), index=[0, 1, 1, 2]),
    pd.Series(np.arange(2), index=[0, 2]),
)
# ValueError: cannot reindex from a duplicate axis

pd.crosstab(
    pd.Series(np.arange(4), index=[0, 1, 1, 2]),
    pd.Series(np.arange(2), index=[3, 4]),
)
# Empty DataFrame
# Columns: []
# Index: []
```

<details>
<summary> Examples with pd.concat </summary>

This had the same behaviour despite not using `.get_indexer` directly in 1.1.5.

```python
pd.concat(
    [
        pd.Series(np.arange(4), index=[0, 1, 1, 2]),
        pd.Series(np.arange(2), index=[0, 2]),
    ],
    axis=1,
    join="inner"
)
# ValueError: cannot reindex from a duplicate axis

pd.concat(
    [
        pd.Series(np.arange(4), index=[0, 1, 1, 2]),
        pd.Series(np.arange(2), index=[3, 4]),
    ],
    axis=1,
    join="inner"
)
# Empty DataFrame
# Columns: [0, 1]
# Index: []
```

</details>

ivirshup replied:
Opened a more general issue for this: #38797

ivirshup committed:
Move dataframe def out of error check

ivirshup committed:
Formatting error

ivirshup committed:
Test allowing duplicate indices if they aren't in intersection (xfail)

ivirshup committed:
Better error messages + organization for duplicate errors

pep8speaks commented:
Hello @ivirshup! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/reshape/concat/test_concat.py`](https://github.com/pandas-dev/pandas/blob/ca316ce5fb5dc3251cfb21679d37febf642719a9/pandas/tests/reshape/concat/test_concat.py):

> [Line 470:89](https://github.com/pandas-dev/pandas/blob/ca316ce5fb5dc3251cfb21679d37febf642719a9/pandas/tests/reshape/concat/test_concat.py#L470): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 88 characters)

ivirshup committed:
Fixes for combining axes when non-intersecting labels are duplicated

This fixes a previous issue where `get_objs_combined_axis` would not throw and error for duplicated indices in all cases.
Now we check to make sure that the duplicated values don't show up in the intersection.
This allows safe indexing when there are duplicated values, but they don't show up in the intersection.

ivirshup commented:
Should the errors here be `DuplicateLabelError`s?

jreback commented on pandas/tests/indexes/test_setops.py:
```diff
@@ -463,3 +466,33 @@ def test_setop_with_categorical(index, sort, method):
     tm.assert_index_equal(result, expected)
+
+
+@pytest.mark.parametrize("index_maker", tm.index_subclass_makers_generator())
```
use the index fixture instead

ivirshup replied:
How should I handle cases where that fixture gives inappropriate input? 

Should I define a separate fixture for this? Can I do something like a [`hypothesis.assume`](https://hypothesis.readthedocs.io/en/latest/details.html#making-assumptions) to make sure some of the cases don't get through and aren't marked as skipped?

The cases which won't work here are at least:

* `"empty"`
* `"repeats"`

jbrockmendel replied:
its kind of ugly, but in many of these tests we do something like

```
def test_foo(index_fixture):
    if whatever(index_fixture):
        pytest.skip()
```

jbrockmendel replied:
(id be psyched to see a hypothesis-based implementation of some of our tests, need a way to generate any-valid-index)

ivirshup replied:
I've switched to the fixture, but it adds a lot of skipped tests. I would be nice if I could filter this before collection, or use a fixture where I can specify some features of the construction (e.g. values are unique and it has a length of `n`)

jreback commented on pandas/tests/indexes/test_setops.py:
```diff
@@ -463,3 +466,33 @@ def test_setop_with_categorical(index, sort, method):
+
+    result = get_objs_combined_axis(series, intersect=True)
+
+    tm.assert_index_equal(full[[0, 2]], result, check_order=False)
```
always use 
result = 
expected = 
tm.assert_index(result, expected)

why is check_order=False?

ivirshup replied:
`check_order` is false because the `MultiIndex` comes back in a different order.

More conceptually, I don't think this function promises any ordering unless you pass sort.

jreback commented on pandas/tests/indexes/test_setops.py:
```diff
@@ -463,3 +466,33 @@ def test_setop_with_categorical(index, sort, method):
+    tm.assert_index_equal(full[[0, 2]], result, check_order=False)
+
+
+@pytest.mark.parametrize("index_maker", tm.index_subclass_makers_generator())
```
same

jreback commented on pandas/tests/indexes/test_setops.py:
```diff
@@ -2,10 +2,13 @@
 The tests in this package are to ensure the proper resultant dtypes of
 set operations.
 """
+from sys import intern
```
what is this for?

ivirshup replied:
I think this is vscode being overzealous with it's auto imports, I'll remove it.

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -445,6 +446,54 @@ def test_concat_ordered_dict(self):
         tm.assert_series_equal(result, expected)
 
 
+@pytest.mark.parametrize("index_maker", tm.index_subclass_makers_generator())
```
same

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -445,6 +446,54 @@ def test_concat_ordered_dict(self):
+@pytest.mark.parametrize("join", ["inner", "outer"])
+def test_concat_duplicates_error(index_maker, join):
+    # if index_maker is tm.makeMultiIndex:
+    # TODO: This generator only makes Indexs of size 4
```
what is this about?

ivirshup replied:
https://github.com/pandas-dev/pandas/blob/0976c4ca0276a333346619c5a9e3c92631ec7006/pandas/_testing/__init__.py#L491-L492

This function only makes a MultiIndex with length 4, ignoring the `k` parameter

ivirshup replied:
#38795

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -445,6 +446,54 @@ def test_concat_ordered_dict(self):
+
+
+@pytest.mark.parametrize("index_maker", tm.index_subclass_makers_generator())
+@pytest.mark.xfail(reason="Not implemented")
```
is there an issue for this? what is this case?

ivirshup replied:
I was expecting this to be allowed:

```python
import pandas as pd

result = pd.concat(
    [
        pd.Series(0, index=[0, 0, 1, 2]),
        pd.Series(1, index=[1, 2]),
    ],
    join="inner",
)
expected = pd.DataFrame({0: [0, 0], 1: [1, 1]}, index=[1, 2])

pd.testing.assert_frame_equal(result, expected)
```

Because the intersection of those indices is well defined. However, it turns out this does not work, and also doesn't work in 1.1.5. I sort of opened this issue here: #38773, but that was a more low-level issue.

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -445,6 +446,54 @@ def test_concat_ordered_dict(self):
+@pytest.mark.parametrize("index_maker", tm.index_subclass_makers_generator())
+@pytest.mark.parametrize("join", ["inner", "outer"])
+def test_concat_duplicates_error(index_maker, join):
+    # if index_maker is tm.makeMultiIndex:
```
add the issue number as a comment in all of the new tests

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -445,6 +446,54 @@ def test_concat_ordered_dict(self):
+
+    tm.assert_frame_equal(
+        result,
+        pd.DataFrame(
```
same as above

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -135,13 +135,20 @@ def _get_combined_index(
     if len(indexes) == 0:
         index = Index([])
-    elif len(indexes) == 1:
+    elif len(indexes) == 1 or all_indexes_same(indexes):
```
add some comments here as it is non-obvious what is happening

ivirshup replied:
How about this?

```suggestion
    # if unique by id or unique by value
    elif len(indexes) == 1 or all_indexes_same(indexes):
```

jreback reviewed (changes requested):
cc @jbrockmendel

jbrockmendel commented:
> because it turns out union and intersection act differently for different index types

@phofl has been working on this recently, might shed some light on the intended behavior medium-long term

ivirshup committed:
Fixes from review and CI

phofl commented:
``union`` and ``intersection`` should both be unique, but as you mentioned this may not be the case for all Index types right now. 

Additionally to your issue #38623 also is problematic. Do you have a list of Index classes where this is not the case for union?

phofl commented:
Put up #38834 to fix the IntervalIndex.intersection bug

ivirshup commented:
> Do you have a list of Index classes where this is not the case for union?

@phofl, let me know if there is an open issue I should post this to. It seems like it's all of them, (for `Index` it depends on the order):

<details>
<summary> Code used to check </summary>

```python
import pandas as pd
import numpy as np

import pandas._testing as tm

index_makers = [
    tm.makeStringIndex,
    tm.makeIntIndex,
    tm.makeUIntIndex,
    tm.makeFloatIndex,
    tm.makeDateIndex,
    tm.makeTimedeltaIndex,
    tm.makePeriodIndex,
    tm.makeMultiIndex,
    tm.makeBoolIndex,
]

records = []

union = lambda x, y: x.union(y)
intersection = lambda x, y: x.intersection(y)

# union = lambda x, y: pd.core.indexes.api._get_combined_index([x, y], intersect=False)
# intersection = lambda x, y: pd.core.indexes.api._get_combined_index([x, y], intersect=True)

for index_maker in index_makers:
    idx1 = index_maker(k=4)
    idx2 = idx1[np.array([0, 0, 1, 2, 3])]
    assert type(idx1) == type(idx2)
    rec = {"index_type": type(idx1).__name__, "index_generator": index_maker.__name__}
    
    # idx_intersect = intersection(idx1, idx2)
    # rec["intersect_len"] = len(idx_intersect)
    # rec["intersect_commutative"] = idx_intersect.sort_values().equals(intersection(idx2, idx1).sort_values())

    idx_union = union(idx1, idx2)
    rec["union_len"] = len(idx_union)
    rec["union_commutative"] = idx_union.sort_values().equals(union(idx2, idx1).sort_values())

    rec["unique_len"] = len(idx_union.unique())
    records.append(rec)

pd.DataFrame.from_records(records)
```

</details>

| index_type     | index_generator    |   union_len | union_commutative   |   unique_len |
|:---------------|:-------------------|------------:|:--------------------|-------------:|
| Index          | makeStringIndex    |           4 | False               |            4 |
| Int64Index     | makeIntIndex       |           5 | True                |            4 |
| UInt64Index    | makeUIntIndex      |           5 | True                |            4 |
| Float64Index   | makeFloatIndex     |           5 | True                |            4 |
| DatetimeIndex  | makeDateIndex      |           5 | True                |            4 |
| TimedeltaIndex | makeTimedeltaIndex |           5 | True                |            4 |
| PeriodIndex    | makePeriodIndex    |           5 | True                |            4 |
| MultiIndex     | makeMultiIndex     |           4 | True                |            4 |
| Index          | makeBoolIndex      |           4 | False               |            2 |


This is using `1.3.0.dev0+218.g3a066feb08`

phofl commented:
Sorry I said something wrong, intersection should be unique, union not.
#36299 will make this consistent for every class except MultiIndex. Will look into this

ivirshup commented:
Why wouldn't unions have unique values?

ivirshup commented:
Currently running into issues with intersections of timestamp objects. They seem to be losing their `offset` field sometimes, not sure what's up with this yet.

phofl commented:
Among others the union case is discussed here: #31326

I think a logical interpretation of a union is, that the union is the smallest index, so that both indexes are part of this index. 

Currently the union is handled differently based on if the indexes are monotonic or not. My pr aims to handle both cases the same and resort afterwards, this would leave us there with more consistency. The unwanted duplicates described above can be solved thereafter I think.

simonjayhawkins commented:
This PR closes #36263 which is milestoned 1.2.1 whereas this PR is milestoned 1.3. what's the best way to resolve this inconsistency?

jreback commented:
this for 1.3

simonjayhawkins commented:
@ivirshup closing as stale. ping if you want to continue.
